src/decode.ts

```diff
@@ -0,0 +1,83 @@
+import { Uint8ArrayList } from 'uint8arraylist'
+import errcode from 'err-code'
+import { FrameHeader, FrameType, HEADER_LENGTH, YAMUX_VERSION } from './frame.js'
+import { ERR_DECODE_INVALID_VERSION } from './constants.js'
+import type { Source } from 'it-stream-types'
+
+/**
+ * Decode a header from the front of a buffer
+ *
+ * @param buffer - Assumed to have enough bytes for a header
+ */
+export function decodeHeader (buffer: Uint8ArrayList): FrameHeader {
+  if (buffer.get(0) !== YAMUX_VERSION) {
+    throw errcode(new Error('Invalid frame version'), ERR_DECODE_INVALID_VERSION)
+  }
+  return {
+    type: buffer.getUint8(1),
+    flag: buffer.getUint16(2, false),
+    streamID: buffer.getUint32(4, false),
+    length: buffer.getUint32(8, false)
+  }
+}
+
+export class Decoder {
+  /** Buffer for in-progress frames */
+  private readonly buffer: Uint8ArrayList
+  private readonly source: Source<Uint8Array>
+
+  constructor (source: Source<Uint8Array>) {
+    this.source = source
+    this.buffer = new Uint8ArrayList()
+  }
+
+  async * emitFrames (): AsyncGenerator<{header: FrameHeader, readData?: () => Promise<Uint8Array>}> {
+    for await (const chunk of this.source) {
+      this.buffer.append(chunk)
+
+      // Loop to consume as many bytes from the buffer as possible
+      // Eg: when a single chunk contains several frames
+      while (true) {
+        const header = this.readHeader()
+        if (header === undefined) {
+          break
+        }
+
+        const { type, length } = header
+        if (type === FrameType.Data) {
+          // If this is a data frame, the frame body must still be read
+          // `readData` must be called before the next iteration here
+          yield {
+            header,
+            readData: async () => await this.readBytes(length)
```

Comment:
why don't we yield the data here? if `readData()` is not called and the next chunk comes, will it try to parse another header unintentionally?

Reply:
It really simplifies the decoder to do this, separation of concerns and all that.
We may or may not read the body of the frame based on the stream, which may or may not exist at this point.

Alternative to this is to have the decoder know about streams, handle stream creation, expose internal stream state, etc.
Or to always read the body of the frame, which opens us up to attacks.

In practice, the header and `readData` are always handled before decoding the next frame (or a fatal protocol error occurs and the muxer sink is stopped)
But we can add some sanity check logic here to blow up / throw an error if `readData` isn't called after the next iteration.

Reply:
Without intimate logic of this usage in practice here I see a closure and extra promise so I raise same concerns as @tuyennhv . How often is this yield hit? Once per stream, packet, etc

Reply:
Once per packet.
Tbh concern here should be balanced against data that this is a hot path. In absence of that, this pattern serves its purpose well: ability to decouple decoding (with conditional reading of data) from stream lifecycle management from stream internal state. Happy to discuss alternatives and refactor if its shown to be an issue.

Reply:
Added some benchmarks. Performance is comparable to mplex.
```
  comparison benchmark
    ✔ yamux send and receive 1 0.0625KB chunks                            600.6728 ops/s    1.664800 ms/op        -        614 runs   1.80 s
    ✔ yamux send and receive 1 1KB chunks                                 602.9265 ops/s    1.658577 ms/op        -        830 runs   2.23 s
    ✔ yamux send and receive 1 64KB chunks                                640.3221 ops/s    1.561714 ms/op        -        396 runs   1.31 s
    ✔ yamux send and receive 1 1024KB chunks                              397.2168 ops/s    2.517517 ms/op        -        744 runs   2.58 s
    ✔ yamux send and receive 1000 0.0625KB chunks                         87.69342 ops/s    11.40336 ms/op        -         28 runs  0.832 s
    ✔ yamux send and receive 1000 1KB chunks                              60.79514 ops/s    16.44868 ms/op        -        125 runs   2.59 s
    ✔ yamux send and receive 1000 64KB chunks                             16.95211 ops/s    58.98972 ms/op        -         17 runs   1.54 s
    ✔ yamux send and receive 1000 1024KB chunks                           1.351564 ops/s    739.8836 ms/op        -         14 runs   11.1 s
    ✔ mplex send and receive 1 0.0625KB chunks                            568.3042 ops/s    1.759621 ms/op        -        205 runs  0.973 s
    ✔ mplex send and receive 1 1KB chunks                                 578.1052 ops/s    1.729789 ms/op        -        725 runs   2.01 s
    ✔ mplex send and receive 1 64KB chunks                                583.7483 ops/s    1.713067 ms/op        -        956 runs   2.40 s
    ✔ mplex send and receive 1 1024KB chunks                              510.3176 ops/s    1.959564 ms/op        -        530 runs   1.66 s
    ✔ mplex send and receive 1000 0.0625KB chunks                         90.75580 ops/s    11.01858 ms/op        -         41 runs  0.973 s
    ✔ mplex send and receive 1000 1KB chunks                              72.11694 ops/s    13.86637 ms/op        -        146 runs   2.55 s
    ✔ mplex send and receive 1000 64KB chunks                             14.81654 ops/s    67.49215 ms/op        -          7 runs   1.03 s
    ✔ mplex send and receive 1000 1024KB chunks                          0.7591994 ops/s    1.317177  s/op        -         12 runs   17.0 s
```